src/fastalloc/lru.rs

```diff
@@ -0,0 +1,240 @@
+use alloc::vec::Vec;
+use alloc::vec;
+use hashbrown::HashSet;
+use core::{fmt, ops::IndexMut};
+use std::{ops::Index, print};
+use crate::{PReg, PRegSet, RegClass};
+
+use std::{println, format};
+
+/// A least-recently-used cache organized as a linked list based on a vector.
+pub struct Lru {
+    /// The list of node information.
+    ///
+    /// Each node corresponds to a physical register.
+    /// The index of a node is the `address` from the perspective of the linked list.
+    pub data: Vec<LruNode>,
+    /// Index of the most recently used register.
+    pub head: usize,
+    /// Class of registers in the cache.
+    pub regclass: RegClass,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct LruNode {
+    /// The previous physical register in the list.
+    pub prev: usize,
+    /// The next physical register in the list.
+    pub next: usize,
+}
+
+impl Lru {
+    pub fn new(regclass: RegClass, regs: &[PReg]) -> Self {
+        let mut data = vec![LruNode { prev: usize::MAX, next: usize::MAX }; PReg::MAX + 1];
+        let no_of_regs = regs.len();
+        for i in 0..no_of_regs {
+            let (reg, prev_reg, next_reg) = (
+                regs[i],
+                regs[i.checked_sub(1).unwrap_or(no_of_regs - 1)],
+                regs[if i >= no_of_regs - 1 { 0 } else { i + 1 }]
+            );
+            data[reg.hw_enc()].prev = prev_reg.hw_enc();
+            data[reg.hw_enc()].next = next_reg.hw_enc();
+        }
+        Self {
+            head: if regs.is_empty() { usize::MAX } else { regs[0].hw_enc() },
+            data,
+            regclass,
+        }
+    }
+
+    /// Marks the physical register `i` as the most recently used
+    /// and sets `vreg` as the virtual register it contains.
+    pub fn poke(&mut self, preg: PReg) {
+        trace!("Before poking: {:?} LRU. head: {:?}, Actual data: {:?}", self.regclass, self.head, self.data);
+        trace!("About to poke {:?} in {:?} LRU", preg, self.regclass);
+        let prev_newest = self.head;
+        let i = preg.hw_enc();
+        if i == prev_newest {
+            return;
+        }
+        if self.data[prev_newest].prev != i {
+            self.remove(i);
+            self.insert_before(i, self.head);
+        }
+        self.head = i;
+        trace!("Poked {:?} in {:?} LRU", preg, self.regclass);
+        self.check_for_cycle();
```

Comment:
This could probably be annotated with `#[cfg(debug_assertions)]` -- that will enable this call when we have assertions compiled in, e.g. for fuzzing or in a general debug build, but avoid it in release builds where performance matters. I suspect that this is probably a reasonably significant slowdown right now if it's running in all your performance tests!

(likewise below too)